Documentation/proposals/app-level-api.md

```diff
@@ -0,0 +1,234 @@
+and then the prepared-app is injected via the entrypoint.
+
+```bash
+rkt app add <pod-uuid> --app=<app-name> <image-name/hash/address/registry-URL> <arguments>
+```
+
+**Note:** Not every pod will be injectable and it will be enabled through an option.
+
+## rkt app rm
+Removes an application image into a running pod, meaning that the leftover resources are removed from the pod.
+
+```bash
+rkt app rm <pod-uuid> --app=<app-name> <arguments>
+```
+
+**Note:** when a pod becomes empty, currently it will terminate, but this
+proposal will introduce a `--mutable` or `--allow-empty` or `--dumb` flag to be
+used when starting pods, so that the lifecycle management of the pod is left to
+the user. Meaning that empty pods won't be terminated.
+
+### Resources Leftover by an Application in Default Stage1 Flavor
+- Rootfs (e.g. `/opt/stage/<app-name>`)
+- Mounts from volumes (e.g. `/opt/stage/<app-name>/<volume-name>`)
+- Mounts related to rkt operations (e.g. `/opt/stage/<app-name>/dev/null`)
+- Systemd service files (e.g. `<app-name>.service` and `reaper-<app-name>.service`)
+- Miscellaneous files (e.g. `/rkt/<app-name>.env`, `/rkt/status...`)
+
+## rkt app start
+Starts an application that was injected.
+This operation is idempotent.
+
+```bash
+rkt app start <pod-uuid> --app=<app-name> <arguments>
+```
+
+## rkt app stop
+Stops a running application gracefully (grace is defined in the `app stop` entrypoint section).
+
+```bash
+rkt app stop <pod-uuid> --app=<app-name>
+```
+
+## rkt app list
+Lists the applications that are inside a pod, running or stopped.
+
+```bash
+rkt app list <pod-uuid> <arguments>
+```
+
+**Note:** List should consist of an app specifier and status at the very least,
+the rest is up for discussions.
+
+## rkt app status
+Returns the execution status of application inside a pod.
+
+```bash
+rkt app status <pod-uuid> --app=<app-name> <arguments>
+```
+
+The returned status information for an example mysql application would contain
+the following details (output format is up for discussion):
+
+```go
+type AppStatus struct {
+	Name       string
+	State      AppState
+	CreatedAt  time.Time
+	StartedAt  time.Time
+	FinishedAt time.Time
+	ExitCode   int64 
+}
+```
+
+## rkt app exec
+Executes a command inside an application.
+
+```bash
+rkt app exec <pod-uuid> --app=<app-name> <arguments> -- <command> <command-arguments>
+```
+
+# Entrypoints
+
+In order to facilitate the app-level operations API, 4 new entrypoints are introduced.
+
+For example, the `app start` entrypoint in default stage1 flavor sends a `start` signal
+(similar to `systemctl start`) for the service files of the application to the systemd which is the stage1.
+
+## rkt app add
+
+`coreos.com/rkt/stage1/app/add`
+
+1. Receive the rootfs, the manifest and the volumes of the prepared app.
+2. `coreos.com/rkt/stage1/add` entrypoint is resolved via annotations found within `/var/lib/rkt/pods/run/$uuid/stage1/manifest`.
```

Comment:
I won't pass the app manifest to an pod but I'll maintain the current implementation where only the pod manifest is used (since you don't need the full app manifest but just the app part and this one is copied from the app manifest to the pod manifest by the stage0 prepare).

Also if a draft I think this is an important point to define and, as suggested by alban [here](https://github.com/coreos/rkt/pull/2932#discussion_r73340628) I'll go for updating (atomically) the pod manifest adding/removing the apps.

(This will also make things like #3039 working without making this proposal appc only)